### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by alessiapacca** [OndrejTexler/Few-Shot-Patch-Based-Training#8]
Hey! 
One question:

With your pretrained weights, can I transfer the style from my image to another by putting:
- the image with the style I want in the folder data_root
- the image I want to transfer the style to in the output folder? 
? 

I don't know if this is possible, but I am asking it because on the presentation videos (https://ondrejtexler.github.io/patch-based_training/) I saw that there was a live style transfer so I guess the network was not retrained, and that it needed only one style example frame. 

in case I wanted to retrain the net, is there a chance I can use a previous version of pytorch? I have cuda 10.0 which is not compatible with pytorch 1.6.0. I actually tried with pytorch 1.2.0 and the training never ever starts, don't know why. 
The stage thing is that by trying on colab it also never starts.

**Comment by alessiapacca** [comment]
or, if i want to re-train the net, should I just have a dataset of images with the style that I want?

**Comment by OndrejTexler** [comment]
Hello. Thanks for the question!

The pre-trained weights will properly work only on the image(s) it was trained on (or on very similar images). If you download _TEST_DATA.zip_ and _pre-trained-models.zip_, and take, for example, the model `_pre-trained-models/Lynx/model_00010.pth`. This model will work best on images from the sequence `TEST_DATA/Lynx_gen/input_filtered`.

To run it on your data, do not use pre-trained weights, you have to train your weights from scratch (but that should not be a problem and it does not require any large GPU servers, just one consumer-grade/laptop NVidia GPU is fine). In Real-Time Live Video (https://www.youtube.com/embed/DDlYIzhoXfI?start=610), we actually trained the network from scratch using one exemplar style within 6 minutes of the real-time live show. Watch it :-) it is explained there and it is really as simple as the explanation says, there is nothing hidden.

I am quite certain that you can use PyTorch 1.2.0 (maybe with some slight modifications). It probably won't be possible to load pre-trained .pth files using the older version of PyTorch, but to train your own weights, it should not be a problem. 

Simply, download TEST_DATA.zip and follow instructions here https://github.com/OndrejTexler/Few-Shot-Patch-Based-Training#train

I hope this helps!

**Comment by alessiapacca** [comment]
Thank you so much for your answer @OndrejTexler, you were super clear!
The problem I had was the mask image I was using that apparently didn't work well. I substituted that one and the training was ok.

I have one more question for you: in case I wanted to make the training more robust (e.g train with more than only one subject for example), can I put more than one picture in the `input_filtered ` folder (pictures of several different subjects) with the corresponding stylized images into the `output` folder? The style would be the same between the pictures, but maybe if trained like this the method could generalize to more than one subject at inference time? 

Let me know!

**Comment by OndrejTexler** [comment]
Exactly, just put multiple pictures to folders `input_filtered`, `output`, and do not forget the `mask` folder. And of course, images in these three folders have to have the same names. In the [TEST_DATA.zip](https://drive.google.com/file/d/1EscSNFg4ILpB7dxr-zYw_UdOILLmDlRj/view), see `Zuzka2_train`, there are 4 training samples. Do your experiments and see how it works. In our experiments, we tried multiple images of the same subject, e.g., photos of a face from different angles, and it works well. But I am not sure whether it will work well on multiple images of completely different people - try it :-)

If you want to experiment with the generalization even more, see configuration file https://github.com/OndrejTexler/Few-Shot-Patch-Based-Training/blob/master/_config/reference_P_disco1010.yaml#L67. In parameter `dir_x1` you can specify another input folder with images, it can be, for instance, segmentation of a face, or any other "guidance channel" you thing might help the network. Again, check `Zuzka2_train`, there is an additional input folder called `input_gdisko_gauss_r10_s10` (the folder with the same name is expected to be in `Zuzka2_gen`)

**Comment by alessiapacca** [comment]
I tried to do a training with 24 different faces in it, initially it started learning in a certain way, then at some point it introduces artifacts in the outputs. I guess it's because it is not really able to understand that there is a common style in all the different images?

**Comment by OndrejTexler** [comment]
It is really hard to tell what is the reason for artifacts. When we were developing the network, we were seeing some artifacts (usually large regions of uniform color not present anywhere in the style). But these artifacts were only in some epochs, and it usually helped to train for a few more seconds/minutes. However, you are trying to use the network in a slightly different scenario than we tuned the network for. 

We tuned our network to work in the scenario described in the paper, i.e., 
- using one or a few (<5) stylized examples
- train quickly, <1 minute
- real-time inference
- sequences we used were usually between 512x512 to 1024x1024 px

If you are using the network in a little bit different scenario, you might want to tune some hyperparameters. See the list of hyperparameters here https://github.com/OndrejTexler/Few-Shot-Patch-Based-Training/blob/master/_config/reference_P.yaml

Hyperparameters I suggest you to try to tune:
- Line 9 filters, if you think the network does not have the capacity to learn "everything" try, for example, double all these values 
- Line 8 resnet_blocks. By increasing resnet_blocks, for example to values 8-15, you also increase the network's capacity and its receptive field (although what the network sees is limited by patch_size)
- Line 17 and Line 34, you might try to set lr (learning rate) to, for example, 0.0002, it might ensure more stable training
- Line 66 patch_size, try to increase this number so that the network sees larger structures
- Line 51 batch_size, you might experiment with this number as well (if you increase patch_size and do not have enough GPU memory, decrease batch_size accordingly)

See the paper (or dig into the code) for details about these parameters :-) 

By tuning the hyperparameters you might end up with slower training, but hopefully, it will yield better results.

**Comment by alessiapacca** [comment]
@OndrejTexler 
Thanks so much for your exhaustive answer!
Is there a chance you still have the code you used to do the grid search for the hyperparameters?

**Comment by OndrejTexler** [comment]
I would not call it "code", it was more like a simple script :-)

In your case, the easiest would be to prepare a couple of `reference_P.yaml` config files and run `train.py` in a batch. Then check the results and visually verify whether it works or not. (to make your experiments easier you can reduce the number of .yaml files that you will need by exposing some parameters, that are defined in .yaml file, into the argument list of `train.py`, see how variables from .yaml file are read in the `train.py`, and override them)

I really hope this will work for you and you will make some interesting findings!
